File: sdks/java/auth/src/main/java/com/ouratoolkit/auth/TokenStore.java
package com.ouratoolkit.auth;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.nio.file.attribute.FileAttribute;
import java.nio.file.attribute.PosixFilePermission;
import java.nio.file.attribute.PosixFilePermissions;
import java.util.Locale;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Semaphore;
import java.util.function.Function;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.cfg.CoercionAction;
import com.fasterxml.jackson.databind.cfg.CoercionInputShape;
import com.fasterxml.jackson.databind.type.LogicalType;

/**
 * Persistent credential store at the fixed, invocation-independent per-platform path:
 * {@code $XDG_CONFIG_HOME/oura-toolkit/} (falling back to
 * {@code $HOME/.config/oura-toolkit/}) on Unix/macOS, {@code %LOCALAPPDATA%\oura-toolkit\}
 * on Windows. This is the SAME directory the Rust CLI/MCP server uses — the two runtimes
 * share one store, so schemas and paths are locked (CLAUDE.md → AUTH).
 *
 * <p>Two records live in the store dir:
 * <ul>
 *   <li><b>{@code credentials.json}</b> — {@link ClientCredentials} (the user's own OAuth
 *       app id + secret). Exists from {@code oura auth setup} onward.</li>
 *   <li><b>{@code tokens.json}</b> — {@link Tokens} (access/refresh/expiry/scope). Exists
 *       only after a successful login; rewritten on every refresh rotation.</li>
 * </ul>
 *
 * <p><b>File hygiene:</b> on POSIX filesystems the records are written {@code 0600} and
 * the directory {@code 0700}, via a uniquely named temp file + atomic move (two
 * concurrent writers can never truncate each other's in-flight temp file; the atomic
 * rename makes the outcome last-writer-wins, never a corrupt mix). On Windows those
 * permission calls are skipped — protection relies on {@code %LOCALAPPDATA%}'s inherited
 * user-private ACLs (Local, not Roaming, deliberately: roaming profiles sync
 * {@code %APPDATA%} off the machine).
 *
 * <p><b>Cross-process coordination:</b> {@link #lockExclusive()} takes an exclusive lock
 * on a {@code .lock} file in the store dir; {@link TokenManager} holds it across its
 * reload → refresh → persist critical section.
 *
 * <p><b>IMPORTANT cross-runtime caveat:</b> on Linux, {@code FileChannel.lock} acquires a
 * POSIX record lock ({@code fcntl}), which does NOT interact with the {@code flock} the
 * Rust CLI/MCP server takes on the same {@code .lock} file — a Java process and a Rust
 * process can both believe they hold "the" lock. The lock still serializes Java processes
 * (and, via a JVM-wide per-path mutex, threads within one JVM — plain
 * {@code FileChannel.lock} would throw {@code OverlappingFileLockException} instead of
 * blocking for same-JVM contention). Across runtimes, the guarantee is the
 * <b>reload-and-adopt + 400-reload-retry protocol</b> in {@link TokenManager}, which is
 * exactly the arm the Rust implementation provides for uncoordinated writers.
 */
public final class TokenStore {

    /** The locked config-directory name (CLAUDE.md → NAMING). */
    public static final String APP_DIR_NAME = "oura-toolkit";

    private static final Set<PosixFilePermission> OWNER_RW =
            PosixFilePermissions.fromString("rw-------");
    private static final Set<PosixFilePermission> OWNER_RWX =
            PosixFilePermissions.fromString("rwx------");

    /**
     * JVM-wide per-store mutexes backing {@link #lockExclusive()}: java.nio file locks
     * are held on behalf of the whole JVM, so same-JVM contention must be serialized
     * here (see the class doc). Deliberately a NON-reentrant {@code Semaphore(1)}, not a
     * {@code ReentrantLock}: flock-style store locks are not reentrant (a thread
     * re-acquiring would then hit {@code OverlappingFileLockException} at the channel),
     * so a second acquire from the SAME thread reports "held" / blocks like any other
     * contender.
     */
    private static final ConcurrentHashMap<String, Semaphore> JVM_LOCKS =
            new ConcurrentHashMap<>();

    private static final ObjectMapper MAPPER = storeMapper();

    private static ObjectMapper storeMapper() {
        ObjectMapper mapper = new ObjectMapper()
                // Match serde's default: unknown fields are ignored, not fatal — a newer
                // writer adding a field must not brick older readers of the shared store.
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                // A JSON `null` for a primitive field (e.g. "expires_at": null) must be a typed
                // store-format error, NOT silently coerced to 0L (which would masquerade as an
                // already-expired token). serde rejects this; match it.
                .configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, true);
        // A JSON number/boolean where a string field is expected (e.g. "client_id": 7)
        // must be a typed store-format error, NOT silently coerced to "7" — serde rejects
        // wrong-typed fields, and the shared store's wire format is serde's (conformance
        // fixture #58, hostile_store_files/credentials_wrong_type_client_id).
        mapper.coercionConfigFor(LogicalType.Textual)
                .setCoercion(CoercionInputShape.Integer, CoercionAction.Fail)
                .setCoercion(CoercionInputShape.Float, CoercionAction.Fail)
                .setCoercion(CoercionInputShape.Boolean, CoercionAction.Fail);
        return mapper;
    }

    private final Path dir;

    /** Store rooted at an explicit directory (tests, or embedded consumers). */
    public TokenStore(Path dir) {
        this.dir = dir;
    }

    /** Store at the default per-platform config location. */
    public static TokenStore openDefault() throws NoConfigDirException {
        return new TokenStore(configDir(System::getenv, isWindows()));
    }

    /** The store directory. */
    public Path getDir() {
        return dir;
    }

    /** Path of the client-credentials record. */
    public Path credentialsPath() {
        return dir.resolve("credentials.json");
    }

    /** Path of the token record. */
    public Path tokensPath() {
        return dir.resolve("tokens.json");
    }

    /**
     * Load the client credentials, or empty if {@code auth setup} has never run. A record
     * that exists but is malformed (bad JSON, wrong-typed/missing field, or a literal
     * {@code null}) surfaces as {@link StoreException}, never a crash.
     */
    public Optional<ClientCredentials> loadCredentials() throws IOException, StoreException {
        return load(credentialsPath(), ClientCredentials.class);
    }

    /** Persist the client credentials ({@code 0600}, atomic). */
    public void saveCredentials(ClientCredentials credentials) throws IOException {
        save(credentialsPath(), credentials);
    }

    /**
     * Load the tokens, or empty if no login has succeeded yet. A record that exists but is
     * malformed (bad JSON, wrong-typed/missing field, or a literal {@code null}) surfaces
     * as {@link StoreException}, never a crash.
     */
    public Optional<Tokens> loadTokens() throws IOException, StoreException {
        return load(tokensPath(), Tokens.class);
    }

    /**
     * Persist the tokens ({@code 0600}, atomic). Callers refreshing MUST persist the
     * rotated refresh token (Oura invalidates the previous one), and MUST do so under
     * {@link #lockExclusive()} as {@link TokenManager} does, or they can burn a rotation
     * another process just persisted.
     */
    public void saveTokens(Tokens tokens) throws IOException {
        save(tokensPath(), tokens);
    }

    /**
     * Take a BLOCKING exclusive lock on the store; hold the returned guard across a
     * reload → refresh → persist critical section (see {@link TokenManager}). Excludes
     * other threads in this JVM (per-path mutex) and other Java processes (file lock);
     * for the Rust runtime see the class-doc caveat — the reload+retry protocol, not this
     * lock, is the cross-runtime guarantee.
     */
    public StoreLock lockExclusive() throws IOException {
        ensureDir();
        Semaphore jvmLock = jvmLockFor();
        jvmLock.acquireUninterruptibly();
        try {
            FileChannel channel = openLockChannel();
            try {
                FileLock fileLock = channel.lock();
                return new StoreLock(jvmLock, channel, fileLock);
            } catch (IOException | RuntimeException e) {
                channel.close();
                throw e;
            }
        } catch (IOException | RuntimeException e) {
            jvmLock.release();
            throw e;
        }
    }

    /**
     * Non-blocking variant of {@link #lockExclusive()}: empty if another holder (thread
     * or process) currently has it.
     */
    public Optional<StoreLock> tryLockExclusive() throws IOException {
        ensureDir();
        Semaphore jvmLock = jvmLockFor();
        if (!jvmLock.tryAcquire()) {
            return Optional.empty();
        }
        try {
            FileChannel channel = openLockChannel();
            try {
                FileLock fileLock = channel.tryLock();
                if (fileLock == null) {
                    channel.close();
                    jvmLock.release();
                    return Optional.empty();
                }
                return Optional.of(new StoreLock(jvmLock, channel, fileLock));
            } catch (IOException | RuntimeException e) {
                channel.close();
                throw e;
            }
        } catch (IOException | RuntimeException e) {
            jvmLock.release();
            throw e;
        }
    }

    // --- internals ---------------------------------------------------------------------

    private Semaphore jvmLockFor() {
        return JVM_LOCKS.computeIfAbsent(
                dir.toAbsolutePath().normalize().toString(), k -> new Semaphore(1));
    }

    private FileChannel openLockChannel() throws IOException {
        Path lockPath = dir.resolve(".lock");
        if (posixSupported()) {
            return FileChannel.open(
                    lockPath,
                    Set.of(StandardOpenOption.CREATE, StandardOpenOption.WRITE),
                    PosixFilePermissions.asFileAttribute(OWNER_RW));
        }
        return FileChannel.open(lockPath, StandardOpenOption.CREATE, StandardOpenOption.WRITE);
    }

    private <T> Optional<T> load(Path path, Class<T> type)
            throws IOException, StoreException {
        final byte[] bytes;
        try {
            bytes = Files.readAllBytes(path);
        } catch (NoSuchFileException e) {
            return Optional.empty();
        }
        final T value;
        try {
            value = MAPPER.readValue(bytes, type);
        } catch (JsonProcessingException e) {
            // Malformed JSON, a wrong-typed or missing required field, or a JSON null for a
            // primitive (FAIL_ON_NULL_FOR_PRIMITIVES): surface ONE typed store-format error,
            // mirroring the Rust store's typed Serde error. The message names the file but
            // NEVER echoes its bytes (the record holds secrets).
            throw new StoreException(path.getFileName() + " is not a valid store record", e);
        }
        if (value == null) {
            // A literal `null` JSON document deserializes to Java null. Left as
            // Optional.of(null) it would throw an UNCHECKED NullPointerException that
            // escapes callers' IOException handling and crashes raw — surface it typed.
            throw new StoreException(
                    path.getFileName() + " is not a valid store record (JSON null)", null);
        }
        return Optional.of(value);
    }

    private void save(Path path, Object record) throws IOException {
        ensureDir();
        byte[] data = MAPPER.writerWithDefaultPrettyPrinter()
                .writeValueAsBytes(record);
        writeSecure(path, data);
    }

    /**
     * Atomic write with owner-only perms: a UNIQUELY named temp file in the same
     * directory (created {@code 0600} on POSIX), fsync, atomic move into place.
     */
    private void writeSecure(Path path, byte[] data) throws IOException {
        boolean posix = posixSupported();
        Path tmp = posix
                ? Files.createTempFile(dir, ".tmp", null,
                        PosixFilePermissions.asFileAttribute(OWNER_RW))
                : Files.createTempFile(dir, ".tmp", null);
        try {
            try (FileChannel ch = FileChannel.open(tmp, StandardOpenOption.WRITE)) {
                ByteBuffer buf = ByteBuffer.wrap(data);
                while (buf.hasRemaining()) {
                    ch.write(buf);
                }
                ch.force(true);
            }
            Files.move(tmp, path, StandardCopyOption.ATOMIC_MOVE,
                    StandardCopyOption.REPLACE_EXISTING);
            if (posix) {
                Files.setPosixFilePermissions(path, OWNER_RW);
            }
        } catch (IOException | RuntimeException e) {
            Files.deleteIfExists(tmp);
            throw e;
        }
    }

    private void ensureDir() throws IOException {
        Files.createDirectories(dir);
        if (posixSupported()) {
            Files.setPosixFilePermissions(dir, OWNER_RWX);
        }
    }

    private boolean posixSupported() {
        return dir.getFileSystem().supportedFileAttributeViews().contains("posix");
    }

    static boolean isWindows() {
        return System.getProperty("os.name", "").toLowerCase(Locale.ROOT).contains("win");
    }

    /**
     * Resolve the fixed, invocation-independent config dir from an injected env lookup
     * (testable on any OS without mutating global env — the store.rs
     * {@code config_dir_from} pattern):
     *
     * <ul>
     *   <li><b>Unix (incl. macOS):</b> {@code $XDG_CONFIG_HOME/oura-toolkit}, falling
     *       back to {@code $HOME/.config/oura-toolkit}. Locked decision — identical under
     *       every launcher, deliberately NOT {@code ~/Library/Application Support}.</li>
     *   <li><b>Windows:</b> {@code %LOCALAPPDATA%\oura-toolkit}. Local, NOT roaming.</li>
     * </ul>
     *
     * <p>EMPTY and RELATIVE env values are treated as absent (XDG spec; a relative base
     * would make where secrets land depend on the process cwd). Absoluteness is checked
     * per the TARGET platform's syntax (a drive-letter check on the Windows branch), so
     * both branches are unit-testable on every CI OS.
     */
    static Path configDir(Function<String, String> env, boolean windows)
            throws NoConfigDirException {
        if (windows) {
            String local = usable(env.apply("LOCALAPPDATA"), true);
            if (local == null) {
                throw new NoConfigDirException(
                        "could not determine the config directory "
                                + "(%LOCALAPPDATA% unset or not an absolute path)");
            }
            return Paths.get(local, APP_DIR_NAME);
        }
        String xdg = usable(env.apply("XDG_CONFIG_HOME"), false);
        if (xdg != null) {
            return Paths.get(xdg, APP_DIR_NAME);
        }
        String home = usable(env.apply("HOME"), false);
        if (home != null) {
            return Paths.get(home, ".config", APP_DIR_NAME);
        }
        throw new NoConfigDirException(
                "could not determine the config directory "
                        + "($XDG_CONFIG_HOME / $HOME unset or not an absolute path)");
    }

    /** Null unless {@code value} is non-empty and absolute for the target platform. */
    private static String usable(String value, boolean windows) {
        if (value == null || value.isEmpty()) {
            return null;
        }
        boolean absolute = windows
                ? value.matches("^[A-Za-z]:[\\\\/].*") || value.startsWith("\\\\")
                : value.startsWith("/");
        return absolute ? value : null;
    }

    /**
     * An exclusive lock on the store, released on {@link #close()}. Always use
     * try-with-resources; the lock releases the moment this closes.
     */
    public static final class StoreLock implements AutoCloseable {
        private final Semaphore jvmLock;
        private final FileChannel channel;
        private final FileLock fileLock;
        private boolean closed;

        private StoreLock(Semaphore jvmLock, FileChannel channel, FileLock fileLock) {
            this.jvmLock = jvmLock;
            this.channel = channel;
            this.fileLock = fileLock;
        }

        @Override
        public void close() {
            if (closed) {
                return;
            }
            closed = true;
            try {
                fileLock.release();
            } catch (IOException ignored) {
                // Channel close below still drops the OS lock.
            } finally {
                try {
                    channel.close();
                } catch (IOException ignored) {
                    // Nothing actionable; the process-exit fallback releases anyway.
                } finally {
                    jvmLock.release();
                }
            }
        }
    }
}

File: sdks/csharp/auth/src/OuraToolkit.Auth/TokenStore.cs
using System.Text.Json;

namespace OuraToolkit.Auth;

/// <summary>
/// Persistent credential store at the fixed, invocation-independent per-platform path
/// shared with every other oura-toolkit consumer (the Rust CLI, the MCP server, the other
/// language companions): <c>$XDG_CONFIG_HOME/oura-toolkit/</c> (falling back to
/// <c>~/.config/oura-toolkit/</c>) on Unix/macOS, <c>%LOCALAPPDATA%\oura-toolkit\</c> on
/// Windows (Local, not Roaming — roaming profiles sync <c>%APPDATA%</c> off the machine).
/// Empty or RELATIVE environment values are ignored (a relative base would make secret
/// placement depend on the process cwd, breaking invocation independence).
///
/// Two records live in the store dir, with snake_case JSON matching the Rust companion:
/// <list type="bullet">
/// <item><c>credentials.json</c> — <see cref="ClientCredentials"/> (the user's own OAuth
/// app id + secret; exists from <c>oura auth setup</c> onward).</item>
/// <item><c>tokens.json</c> — <see cref="Tokens"/> (access/refresh/expiry/scope; rewritten
/// on every refresh rotation).</item>
/// </list>
///
/// File hygiene: on Unix, records are created 0600 (and the dir 0700) and every write is
/// atomic (uniquely named temp file + rename), so concurrent writers can never mix bytes —
/// last writer wins. On Windows the chmods are no-ops; protection relies on
/// <c>%LOCALAPPDATA%</c>'s user-private ACLs.
///
/// Cross-process coordination: <see cref="AcquireLockAsync"/> takes an exclusive lock on a
/// <c>.lock</c> file in the store dir; <see cref="TokenManager"/> holds it across its
/// reload → refresh → persist critical section. NOTE: whether this lock and the Rust CLI's
/// <c>flock</c> exclude each other is a runtime implementation detail (on Linux .NET
/// happens to map <c>FileShare.None</c> onto <c>flock</c>, but that is NOT a contract) —
/// the cross-runtime guarantee is the PROTOCOL, not the lock: every refresher re-reads the
/// store first, adopts an already-performed rotation instead of re-burning it, and retries
/// a refresh 400 once against freshly reloaded disk state (see <see cref="TokenManager"/>).
/// </summary>
public sealed class TokenStore
{
    /// <summary>
    /// The locked config-directory name (CLAUDE.md → NAMING), identical under every install
    /// path on every platform.
    /// </summary>
    public const string AppDirName = "oura-toolkit";

    /// <summary>Store at the default per-platform config location.</summary>
    public TokenStore()
        : this(ResolveConfigDir(Environment.GetEnvironmentVariable, !IsPosix))
    {
    }

    /// <summary>
    /// True when NOT running on Windows (i.e. a POSIX host whose file modes matter). Carried
    /// as a single guard so the two legs differ only in HOW they answer: net7+ uses
    /// <c>OperatingSystem.IsWindows()</c> — which the platform-compatibility analyzer also
    /// recognizes (via the <c>[UnsupportedOSPlatformGuard("windows")]</c> on this member)
    /// as a guard for the <c>[UnsupportedOSPlatform("windows")]</c> UnixFileMode APIs —
    /// while netstandard2.0 (no <c>OperatingSystem</c> type) uses <c>RuntimeInformation</c>.
    /// </summary>
#if !NETSTANDARD2_0
    [System.Runtime.Versioning.UnsupportedOSPlatformGuard("windows")]
#endif
    private static bool IsPosix =>
#if NETSTANDARD2_0
        !System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(
            System.Runtime.InteropServices.OSPlatform.Windows);
#else
        !OperatingSystem.IsWindows();
#endif

    /// <summary>Store rooted at an explicit directory (used by tests).</summary>
    public TokenStore(string directory) => Directory = directory;

    /// <summary>The store directory.</summary>
    public string Directory { get; }

    /// <summary>Path of the client-credentials record.</summary>
    public string CredentialsPath => Path.Combine(Directory, "credentials.json");

    /// <summary>Path of the token record.</summary>
    public string TokensPath => Path.Combine(Directory, "tokens.json");

    private string LockPath => Path.Combine(Directory, ".lock");

    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    /// <summary>Load the client credentials, or null if <c>auth setup</c> has never run.</summary>
    public ClientCredentials? LoadCredentials() => LoadRecord<ClientCredentials>(CredentialsPath);

    /// <summary>Persist the client credentials (0600 on Unix, atomic).</summary>
    public void SaveCredentials(ClientCredentials credentials)
    {
        EnsureDirectory();
        WriteSecure(CredentialsPath, JsonSerializer.SerializeToUtf8Bytes(credentials, JsonOptions));
    }

    /// <summary>Load the tokens, or null if no login has succeeded yet.</summary>
    public Tokens? LoadTokens() => LoadRecord<Tokens>(TokensPath);

    /// <summary>
    /// Persist the tokens (0600 on Unix, atomic). Callers refreshing MUST persist the
    /// rotated refresh token (Oura invalidates the previous one), and MUST do so under
    /// <see cref="AcquireLockAsync"/> as <see cref="TokenManager"/> does — the atomic write
    /// prevents corruption, not rotation burns.
    /// </summary>
    public void SaveTokens(Tokens tokens)
    {
        EnsureDirectory();
        WriteSecure(TokensPath, JsonSerializer.SerializeToUtf8Bytes(tokens, JsonOptions));
    }

    /// <summary>
    /// Take an exclusive lock on the store, waiting (polling) until the current holder
    /// releases it. Hold the returned handle across a reload → refresh → persist critical
    /// section; dispose to release. The hold duration is bounded by the token endpoint's
    /// hard timeout (<see cref="TokenManager.TokenEndpointTimeout"/>).
    /// </summary>
    public async Task<StoreLock> AcquireLockAsync(CancellationToken cancellationToken = default)
    {
        while (true)
        {
            if (TryAcquireLock() is { } held)
            {
                return held;
            }
            await Task.Delay(TimeSpan.FromMilliseconds(25), cancellationToken).ConfigureAwait(false);
        }
    }

    /// <summary>
    /// Non-blocking variant of <see cref="AcquireLockAsync"/>: null if another holder
    /// currently has the lock.
    /// </summary>
    public StoreLock? TryAcquireLock()
    {
        EnsureDirectory();
        try
        {
            return new StoreLock(OpenExclusiveLock(LockPath));
        }
        catch (IOException)
        {
            // Held by someone else (sharing violation). Genuine I/O failures (permissions,
            // missing dir) also land here in rare cases, but the retry loop's caller keeps
            // a bounded critical section, and EnsureDirectory above rules out the dir.
            return null;
        }
    }

    /// <summary>
    /// Open the <c>.lock</c> file with exclusive sharing (<see cref="FileShare.None"/>), the
    /// primitive the advisory lock is built on. The lock file is secret-free, so its 0600 on
    /// Unix is hygiene, not a security guarantee. net7+: <c>FileStreamOptions.UnixCreateMode</c>
    /// sets 0600 at creation. netstandard2.0: the classic constructor (no mode support), then a
    /// best-effort <c>chmod</c> — a brief non-0600 window is acceptable for a contentless lock.
    /// </summary>
    private static FileStream OpenExclusiveLock(string path)
    {
#if NETSTANDARD2_0
        var stream = new FileStream(path, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None);
        if (IsPosix)
        {
            PosixInterop.TryChmod(path, PosixInterop.Mode0600);
        }
        return stream;
#else
        var options = new FileStreamOptions
        {
            Mode = FileMode.OpenOrCreate,
            Access = FileAccess.ReadWrite,
            // Exclusivity: no other handle (in this or any other process honoring the
            // sharing mode) may hold the file while we do.
            Share = FileShare.None,
        };
        if (IsPosix)
        {
            options.UnixCreateMode = UnixFileMode.UserRead | UnixFileMode.UserWrite;
        }
        return new FileStream(path, options);
#endif
    }

    private T? LoadRecord<T>(string path) where T : class
    {
        byte[] bytes;
        try
        {
            bytes = File.ReadAllBytes(path);
        }
        catch (FileNotFoundException)
        {
            return null;
        }
        catch (DirectoryNotFoundException)
        {
            return null;
        }
        catch (UnauthorizedAccessException e)
        {
            // The record path is not a readable file — most commonly it is itself a directory
            // (File.ReadAllBytes on a directory throws this on Unix). Surface it as a typed
            // store-format error, not a raw IO exception leaking to callers.
            throw new StoreFormatException(path, e);
        }
        try
        {
            return JsonSerializer.Deserialize<T>(bytes)
                ?? throw new StoreFormatException(path, new JsonException("record is JSON null"));
        }
        catch (JsonException e)
        {
            throw new StoreFormatException(path, e);
        }
    }

    private void EnsureDirectory()
    {
        System.IO.Directory.CreateDirectory(Directory);
        if (IsPosix)
        {
            // 0700, every time (CreateDirectory does not re-mode an existing dir).
#if NETSTANDARD2_0
            PosixInterop.TryChmod(Directory, PosixInterop.Mode0700);
#else
            File.SetUnixFileMode(
                Directory,
                UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute);
#endif
        }
    }

    /// <summary>
    /// Atomic write with owner-only perms: a UNIQUELY named temp file in the same directory
    /// (created 0600 on Unix), flushed to disk, then renamed into place. The unique name
    /// means two concurrent writers can never truncate each other's in-flight temp file;
    /// the rename makes the outcome last-writer-wins, never a corrupt mix.
    /// </summary>
    private static void WriteSecure(string path, byte[] data)
    {
        var dir = Path.GetDirectoryName(path)
            ?? throw new InvalidOperationException("store paths always have a parent dir");
        var temp = Path.Combine(dir, $".tmp-{Guid.NewGuid():N}");
        try
        {
            using (var stream = CreateExclusiveOwnerOnly(temp))
            {
                // Write(byte[], int, int) — not the Write(ReadOnlySpan<byte>) overload, which
                // is absent on netstandard2.0; this form is identical on every TFM.
                stream.Write(data, 0, data.Length);
                stream.Flush(flushToDisk: true);
            }
            ReplaceAtomically(temp, path);
        }
        catch
        {
            try
            {
                File.Delete(temp);
            }
            catch (IOException)
            {
                // Best-effort cleanup; the original error is what matters.
            }
            throw;
        }
    }

    /// <summary>
    /// Create a brand-new temp file for writing, owner-only (0600) on Unix, failing if it
    /// somehow already exists. net7+: <c>FileStreamOptions.UnixCreateMode</c> pins the mode at
    /// creation. netstandard2.0: <see cref="PosixInterop.CreateExclusive0600"/> does the same
    /// via <c>open(O_CREAT|O_EXCL, 0600)</c> — NO wider window than the modern leg. On
    /// Windows/.NET Framework, <c>FileMode.CreateNew</c> and the user-private ACLs of
    /// <c>%LOCALAPPDATA%</c> provide the protection.
    /// </summary>
    private static FileStream CreateExclusiveOwnerOnly(string temp)
    {
#if NETSTANDARD2_0
        if (IsPosix)
        {
            return PosixInterop.CreateExclusive0600(temp);
        }
        return new FileStream(temp, FileMode.CreateNew, FileAccess.Write, FileShare.None);
#else
        var options = new FileStreamOptions
        {
            Mode = FileMode.CreateNew,
            Access = FileAccess.Write,
            Share = FileShare.None,
        };
        if (IsPosix)
        {
            options.UnixCreateMode = UnixFileMode.UserRead | UnixFileMode.UserWrite;
        }
        return new FileStream(temp, options);
#endif
    }

    /// <summary>
    /// Atomically move <paramref name="temp"/> onto <paramref name="path"/>, replacing any
    /// existing record last-writer-wins. net5+: <c>File.Move(overwrite: true)</c>.
    /// netstandard2.0 on Unix: <c>rename(2)</c> (atomic replace). netstandard2.0 on
    /// Windows/.NET Framework: the 2-arg <c>File.Move</c> refuses an existing destination, so
    /// delete-then-move — a brief window covered by <c>%LOCALAPPDATA%</c>'s ACLs, the same
    /// place that leg relies on for confidentiality anyway.
    /// </summary>
    private static void ReplaceAtomically(string temp, string path)
    {
#if NETSTANDARD2_0
        if (IsPosix)
        {
            PosixInterop.Rename(temp, path);
            return;
        }
        if (File.Exists(path))
        {
            File.Delete(path);
        }
        File.Move(temp, path);
#else
        File.Move(temp, path, overwrite: true);
#endif
    }

    /// <summary>
    /// Testable core of the default-path resolution (injected env lookup — mirrors the Rust
    /// companion's <c>config_dir_from</c>; no racy global env mutation in tests, and both
    /// platform branches are exercisable everywhere via <paramref name="isWindows"/>).
    ///
    /// Empty and RELATIVE values are treated as absent (XDG spec: relative values should be
    /// ignored) — a relative base would make where secrets land depend on the process cwd.
    /// </summary>
    internal static string ResolveConfigDir(Func<string, string?> env, bool isWindows)
    {
        string? Usable(string key)
        {
            var value = env(key);
            // value is non-null in the true branch (guarded by !IsNullOrEmpty); the `!` is for
            // the netstandard2.0 BCL, whose IsNullOrEmpty lacks [NotNullWhen(false)].
            return !string.IsNullOrEmpty(value) && IsAbsolute(value!, isWindows) ? value : null;
        }

        var separator = isWindows ? '\\' : '/';
        if (isWindows)
        {
            // Local, NOT roaming (%APPDATA%): roaming profiles sync to file servers and
            // profile backups at logoff, which would copy plaintext OAuth secrets off the
            // machine.
            return Usable("LOCALAPPDATA") is { } local
                ? $"{local.TrimEnd(separator)}{separator}{AppDirName}"
                : throw new NoConfigDirException(isWindows: true);
        }
        if (Usable("XDG_CONFIG_HOME") is { } xdg)
        {
            return $"{xdg.TrimEnd(separator)}{separator}{AppDirName}";
        }
        return Usable("HOME") is { } home
            ? $"{home.TrimEnd(separator)}{separator}.config{separator}{AppDirName}"
            : throw new NoConfigDirException(isWindows: false);
    }

    /// <summary>
    /// Platform-aware absolute-path check, independent of the RUNNING platform so both
    /// branches are unit-testable on any CI leg (Path.IsPathRooted answers only for the
    /// current OS). Windows: drive-rooted (<c>C:\</c> or <c>C:/</c>) or UNC (<c>\\</c>);
    /// drive-relative forms like <c>C:foo</c> or bare <c>\foo</c> do not qualify — they
    /// still depend on per-process state. Unix: leading <c>/</c>.
    /// </summary>
    private static bool IsAbsolute(string path, bool isWindows)
    {
        if (!isWindows)
        {
            return path.Length > 0 && path[0] == '/';
        }
        if (path.StartsWith(@"\\", StringComparison.Ordinal))
        {
            return true; // UNC
        }
        return path.Length >= 3
            && IsAsciiLetter(path[0])
            && path[1] == ':'
            && (path[2] == '\\' || path[2] == '/');
    }

    // char.IsAsciiLetter is net7+; this ASCII-only check is TFM-agnostic and equivalent here
    // (drive letters are ASCII).
    private static bool IsAsciiLetter(char c) =>
        (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

/// <summary>
/// An exclusive lock on the store, released on dispose (or process exit). Bind it for the
/// whole critical section — releasing early re-opens the rotation race.
/// </summary>
public sealed class StoreLock : IDisposable
{
    private readonly FileStream _file;

    internal StoreLock(FileStream file) => _file = file;

    /// <summary>Releases the lock.</summary>
    public void Dispose() => _file.Dispose();
}
